engine/src/main/java/org/hibernate/validator/cfg/scriptengine/impl/AbstractCacheableScriptEvaluatorFactory.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.cfg.scriptengine.impl;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import javax.script.ScriptEngineManager;
+import javax.script.ScriptException;
+import javax.validation.ConstraintDeclarationException;
+import javax.validation.ValidationException;
+
+import org.hibernate.validator.cfg.scriptengine.ScriptEvaluator;
+import org.hibernate.validator.cfg.scriptengine.ScriptEvaluatorFactory;
+
+/**
+ * Basic cacheable factory responsible for the creation of {@link ScriptEvaluator}s. This
+ * class is thread-safe. Caches {@code ScriptEvaluator} when they are requested.
+ *
+ * @author Gunnar Morling
+ * @author Kevin Pollet &lt;kevin.pollet@serli.com&gt; (C) 2011 SERLI
+ * @author Marko Bekhta
+ */
+public abstract class AbstractCacheableScriptEvaluatorFactory implements ScriptEvaluatorFactory {
```

Comment:
I don't think we should provide any user-facing implementations (assuming that this is what you intended to do with that `impl` package). We can add a sketch for OSGi to the docs, but I don't feel it's something I'd like to maintain as part of the library proper.